README
Copyright 2007, 2008 Kevin Ryde

This file is part of Gtk2::Ex::NoShrink.

Gtk2::Ex::NoShrink is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation; either version 3, or (at
your option) any later version.

Gtk2::Ex::NoShrink is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License along
with Gtk2::Ex::NoShrink.  If not, see <http://www.gnu.org/licenses/>.




Gtk2::Ex::NoShrink is a simple container widget imposing a "no shrink"
policy on the size requested by its child.  If the child asks to
become smaller then that request is ignored, rather the child's
largest-ever request is used.

The actual size the NoShrink and in turn its child get depends on the
further higher widget parents.  But for forms or tables where size
requests provoke a re-layout the NoShrink can stop pointlessly small
wobbles, or if the child's desired size is data-dependent but too hard
to calculate in advance then it can let the child reach a peak and
then maintain that.

The only prerequisite is Gtk2-Perl (the Gtk2 modules).  The NoShrink
home page is

    http://www.geocities.com/user42_kevin/gtk-ex-noshrink/index.html

Bug reports to the author at user42@zip.com.au please.


Changes
Copyright 2007, 2008, 2009, 2010 Kevin Ryde

This file is part of Gtk2-Ex-NoShrink.

Gtk2-Ex-NoShrink is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation; either version 3, or (at
your option) any later version.

Gtk2-Ex-NoShrink is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License along
with Gtk2-Ex-NoShrink.  If not, see <http://www.gnu.org/licenses/>.


Version 3, February 2010
    - no redraw_on_allocation since NoShrink doesn't draw

Version 2, January 2008
    - lose "-1" stuff for no specified size.

Version 1, January 2008
    - the first version.


Makefile.PL
#!/usr/bin/perl

# Copyright 2007, 2008, 2009, 2010 Kevin Ryde

# This file is part of Gtk2-Ex-NoShrink.
#
# Gtk2-Ex-NoShrink is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 3, or (at your option) any later
# version.
#
# Gtk2-Ex-NoShrink is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#
# You should have received a copy of the GNU General Public License along
# with Gtk2-Ex-NoShrink.  If not, see <http://www.gnu.org/licenses/>.

use 5.008;
use strict;
use warnings;
use ExtUtils::MakeMaker;
use lib 'inc';
use MyMakeMakerExtras;

MyMakeMakerExtras::WriteMakefile
  (NAME         => 'Gtk2-Ex-NoShrink',
   ABSTRACT     => 'Non-shrinking container widget for Gtk.',
   VERSION_FROM => 'lib/Gtk2/Ex/NoShrink.pm',
   AUTHOR       => 'Kevin Ryde <user42@zip.com.au>',
   LICENSE      => 'gpl',
   SIGN         => 1,
   MIN_PERL_VERSION => '5.008',

   PREREQ_PM    => {
                    # part of Glib, used explicitly so name it
                    'Glib::Object::Subclass' => 0,

                    'Gtk2' => 0, # any version should be ok
                   },

   META_MERGE
   => { resources
        => { homepage => 'http://user42.tuxfamily.org/gtk2-ex-noshrink/index.html' },
        optional_features
        => { maximum_tests
             => { description => 'Have "make test" do as much as possible.',
                  requires => { 'Test::NoWarnings' => 0,
                                'Test::Weaken' => 3,
                              },
                },
           },
      },
  );


devel/run.pl
#!/usr/bin/perl

# Copyright 2009 Kevin Ryde

# This file is part of Gtk2-Ex-NoShrink.
#
# Gtk2-Ex-NoShrink is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 3, or (at your option) any later
# version.
#
# Gtk2-Ex-NoShrink is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#
# You should have received a copy of the GNU General Public License along
# with Gtk2-Ex-NoShrink.  If not, see <http://www.gnu.org/licenses/>.


use strict;
use warnings;
use Gtk2 '-init';
use Gtk2::Ex::NoShrink;

use FindBin;
my $progname = $FindBin::Script;

my $toplevel = Gtk2::Window->new('toplevel');
my $screen_width = $toplevel->get_screen->get_width;
$toplevel->set_default_size ($screen_width / 2, -1);
$toplevel->signal_connect (destroy => sub { Gtk2->main_quit; });

#my $layout = Gtk2::Frame->new;
my $layout = Gtk2::Layout->new;
$toplevel->add ($layout);

my $noshrink = Gtk2::Ex::NoShrink->new (shrink_width_factor => 2);
$noshrink->set_border_width (5);
$noshrink->signal_connect (size_allocate => sub {
                             my ($noshrink, $rect) = @_;
                             print "$progname: NoShrink size_allocate ",
                               $rect->x,",",$rect->y,
                                 " ",$rect->width,"x",$rect->height,
                                   "\n";
                           });
$noshrink->signal_connect (expose_event => sub {
                             my ($noshrink, $event) = @_;
                             print "$progname: NoShrink expose ",
                               region_to_string($event->region),"\n";
                             return 0; # Gtk2::EVENT_PROPAGATE
                           });
$layout->add ($noshrink);

my $adj = Gtk2::Adjustment->new (40, 0,$screen_width, 1,10,0);

my $spin = Gtk2::SpinButton->new ($adj, 10, 0);
$spin->set_size_request ($spin->get_value, -1);
$noshrink->add ($spin);
$spin->signal_connect (value_changed => sub {
                         # width from the spinner value
                         $spin->set_size_request ($spin->get_value, -1);
                       });
$spin->signal_connect (size_allocate => sub {
                         my ($noshrink, $rect) = @_;
                         print "$progname: SpinButton size_allocate ",
                           $rect->x,",",$rect->y,
                             " ",$rect->width,"x",$rect->height,
                               "\n";
                       });
# $spin->signal_connect (expose_event => sub {
#                          my ($spin, $event) = @_;
#                          print "$progname: SpinButton expose ",
#                            region_to_string($event->region),"\n";
#                          return 0; # Gtk2::EVENT_PROPAGATE
#                        });

Gtk2->key_snooper_install
  (sub {
     my ($widget, $event) = @_;
     $event->type eq 'key-press' or return 0;  # Gtk2::EVENT_PROPAGATE
     my $key = Gtk2::Gdk->keyval_name ($event->keyval);
     print "$progname: key $key\n";
     
     if ($key eq 'l' && $event->state & 'control-mask') {
       print "$progname: queue_draw\n";
       $toplevel->queue_draw;
       return 1;  # Gtk2::EVENT_STOP
     }
     if ($key eq 'Down' && $event->state & 'control-mask') {
       my $x = $layout->child_get_property($noshrink,'x');
       my $y = $layout->child_get_property($noshrink,'y');
       $y += 10;
       print "$progname: move noshrink to $x,$y\n";
       $layout->move($noshrink, $x, $y);
       return 1;  # Gtk2::EVENT_STOP
     }
     if ($key eq 'Up' && $event->state & 'control-mask') {
       my $x = $layout->child_get_property($noshrink,'x');
       my $y = $layout->child_get_property($noshrink,'y');
       $y -= 10;
       print "$progname: move noshrink to $x,$y\n";
       $layout->move($noshrink, $x, $y);
       return 1;  # Gtk2::EVENT_STOP
     }
     return 0;  # Gtk2::EVENT_PROPAGATE
   });

my $req = $noshrink->size_request;
$layout->set_size_request (-1, $req->height * 3);

$toplevel->show_all;
Gtk2->main;
exit 0;



sub region_to_string {
  my ($region) = @_;
  if (my @rectangles = $region->get_rectangles) {
    return join (' ', map {$_->x.','.$_->y.'='.$_->width.'x'.$_->height}
                 @rectangles);
  } else {
    return 'empty';
  }
}


examples/spin.pl
#!/usr/bin/perl

# Copyright 2007, 2008, 2009 Kevin Ryde

# This file is part of Gtk2-Ex-NoShrink.
#
# Gtk2-Ex-NoShrink is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 3, or (at your option) any later
# version.
#
# Gtk2-Ex-NoShrink is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#
# You should have received a copy of the GNU General Public License along
# with Gtk2-Ex-NoShrink.  If not, see <http://www.gnu.org/licenses/>.


# NoShrink widget with size requested by numeric spinner.
#
# This is a manually operated example.  The desired size of the spinner
# widget is set (set_size_request), in pixels, from the value you enter or
# scroll up or down to.  Then because it's inside a NoShrink container the
# actual request (and hence the allocated size) is subject to that widget's
# no-shrink policy.
#
# The NoShrink has the shrink_width_factor option is set in this example and
# you can see how it allows shrinks of a certain big enough amount.  In this
# case a setting 2 means a factor of 2 (or more) smaller than the current
# recorded peak will be obeyed (and it resets the peak to that new smaller
# size).  So for instance if you run the spinner up to 300 pixels and then
# start running it down again, when it reaches 150 pixel that size will be
# obeyed.
#
# If you uncomment the line "$noshrink = Gtk2::Frame->new;" you can see what
# happens in a GtkFrame container instead of a NoShrink.  A frame simply
# passes its child's size requests upwards, so running the spinner up or
# down in that case is immediately reflected in the spinners size.
#
# Incidentally, everything works the same vertically on requested heights
# (width and height are handled independently).  But for this example the
# spinner looks better getting wider and narrower than taller and shorter.


use strict;
use warnings;
use Gtk2 '-init';
use Gtk2::Ex::NoShrink;

my $toplevel = Gtk2::Window->new('toplevel');
my $screen_width = $toplevel->get_screen->get_width;
$toplevel->set_default_size ($screen_width / 2, -1);
$toplevel->signal_connect (destroy => sub { Gtk2->main_quit; });

my $adj = Gtk2::Adjustment->new (40,            # initial value
                                 0,             # lower
                                 $screen_width, # upper
                                 1,             # step_increment
                                 10,            # page_increment
                                 0);            # page_size not applicable

my $spin = Gtk2::SpinButton->new ($adj, 10, 0);
$spin->set_size_request ($spin->get_value, -1);
$spin->signal_connect (value_changed => sub {
                         # width from the spinner value
                         $spin->set_size_request ($spin->get_value, -1);
                       });

my $noshrink = Gtk2::Ex::NoShrink->new (shrink_width_factor => 2);
#
# try this for plain frame instead of noshrink:
# $noshrink = Gtk2::Frame->new;
#
$noshrink->set_border_width (5);
$noshrink->add ($spin);

my $layout = Gtk2::Layout->new;
$layout->add ($noshrink);
$layout->show_all;
$toplevel->add ($layout);

# lock height of the spinner and any noshrink border width because a layout
# widget doesn't otherwise offer a desired height
my $req = $noshrink->size_request;
$layout->set_size_request (-1, $req->height);

$toplevel->show;
Gtk2->main;
exit 0;


examples/expander.pl
#!/usr/bin/perl

# Copyright 2007, 2008 Kevin Ryde

# This file is part of Gtk2-Ex-NoShrink.
#
# Gtk2-Ex-NoShrink is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 3, or (at your option) any later
# version.
#
# Gtk2-Ex-NoShrink is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#
# You should have received a copy of the GNU General Public License along
# with Gtk2-Ex-NoShrink.  If not, see <http://www.gnu.org/licenses/>.


# Demonstrate NoShrink with a pulsing expanding label.
#
# This is an automated demo of the noshrink policy.  The text of a label
# widget is set to a run of stars "****" growing and shrinking on a timer.
# The changing text makes the label ask to be bigger and smaller, but the
# NoShrink widget keeps it at its peak size.
#
# The label is setup as white text on a black background, and that can be
# seen growing into space provided by a GtkLayout widget (with a white
# background).
#
# The "xalign" property is set to 0 in the label to have the text aligned at
# the left when the space is more than the text needs.  You can change that
# to say 1.0 to have it aligned at the right -- the no-shrink policy remains
# the same.
#
# If you're wondering what the EventBox is for, it just gives a different
# background colour for the label.  A label is a no-window widget (and
# NoShrink is a no-window too) and hence gets the background of its parent,
# put it in something with a window to have a background colour contrasting
# against the layout parent.


use strict;
use warnings;
use Gtk2 '-init';
use Gtk2::Ex::NoShrink;
use Math::Trig;


Gtk2::Rc->parse_string ('
style "White_on_Black" {
    fg[NORMAL]    = { 1.0, 1.0, 1.0 }
    bg[NORMAL]    = { 0, 0, 0 }
    base[NORMAL]  = { 0, 0, 0 }
}
style "Black_on_White" {
    fg[NORMAL]    = { 0, 0, 0 }
    text[NORMAL]  = { 0, 0, 0 }
    bg[NORMAL]    = { 1.0, 1.0, 1.0 }
}
widget "*.GtkLabel"    style "White_on_Black"
widget "*.GtkEventBox" style "White_on_Black"
widget "*.GtkLayout"   style "Black_on_White"
');

my $label = Gtk2::Label->new ('*');
$label->set (xalign => 0);

my $eventbox = Gtk2::EventBox->new;
$eventbox->add ($label);

my $noshrink = Gtk2::Ex::NoShrink->new (child => $eventbox);

my $layout = Gtk2::Layout->new;
$layout->add ($noshrink);
# use label's desired height
my $req = $label->size_request;
$layout->set_size_request (-1, $req->height);

my $toplevel = Gtk2::Window->new('toplevel');
$toplevel->add ($layout);
$toplevel->set_default_size ($toplevel->get_screen->get_width / 2, -1);
$toplevel->signal_connect (destroy => sub { Gtk2->main_quit; });
$toplevel->show_all;

my $x = 0;
sub timer_callback {
  my $len = int ($x/4 + 10 * sin ($x));
  $label->set_text ('*' x $len);
  $x += Math::Trig::pi / 10;
  return 1; # Glib::SOURCE_CONTINUE
}
Glib::Timeout->add (100, \&timer_callback);

Gtk2->main;
exit 0;


lib/Gtk2/Ex/NoShrink.pm
# Copyright 2007, 2008, 2009, 2010 Kevin Ryde

# This file is part of Gtk2-Ex-NoShrink.
#
# Gtk2-Ex-NoShrink is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 3, or (at your option) any later
# version.
#
# Gtk2-Ex-NoShrink is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#
# You should have received a copy of the GNU General Public License along
# with Gtk2-Ex-NoShrink.  If not, see <http://www.gnu.org/licenses/>.

package Gtk2::Ex::NoShrink;
use 5.008;
use strict;
use warnings;
use Gtk2;
use List::Util qw(min max);
use POSIX ();

our $VERSION = 3;

# set this to 1 for some diagnostic prints
use constant DEBUG => 0;

use Glib::Object::Subclass
  'Gtk2::Bin',
  signals => { size_allocate => \&do_size_allocate,
               size_request  => \&do_size_request,
             },
  properties => [Glib::ParamSpec->int
                 ('minimum-width',
                  'minimum-width',
                  '',
                  0, POSIX::INT_MAX(), # range
                  0,                   # default
                  Glib::G_PARAM_READWRITE),
                 Glib::ParamSpec->int
                 ('minimum-height',
                  'minimum-height',
                  '',
                  0, POSIX::INT_MAX(), # range
                  0,                   # default
                  Glib::G_PARAM_READWRITE),

                 Glib::ParamSpec->double
                 ('shrink-width-factor',
                  'shrink-width-factor',
                  '',
                  0, POSIX::DBL_MAX(), # range
                  0,                   # default
                  Glib::G_PARAM_READWRITE),
                 Glib::ParamSpec->double
                 ('shrink-height-factor',
                  'shrink-height-factor',
                  '',
                  0, POSIX::DBL_MAX(), # range
                  0,                   # default
                  Glib::G_PARAM_READWRITE)
                ];

sub INIT_INSTANCE {
  my ($self) = @_;

  # NoShrink doesn't draw into the window
  # (child still gets exposes on any resize though)
  $self->set_redraw_on_allocate(0);

  # per defaults in the ParamSpec's above
  $self->{'minimum_width'}  = 0;
  $self->{'minimum_height'} = 0;
  $self->{'shrink_width_factor'} = 0;
  $self->{'shrink_height_factor'} = 0;
}

sub SET_PROPERTY {
  my ($self, $pspec, $newval) = @_;
  my $pname = $pspec->get_name;
  my $oldval = $self->{$pname};
  $self->{$pname} = $newval;  # per default GET_PROPERTY

  if ($oldval != $newval) {
    $self->queue_resize;
  }
}

# 'size-request' class closure
#
# called by anyone interested in how big we want to be -- ask child and add
# the border width
# 
sub do_size_request {
  my ($self, $req) = @_;

  my $old_min_width  = $self->{'minimum_width'};
  my $old_min_height = $self->{'minimum_height'};
  my $min_width  = $old_min_width;
  my $min_height = $old_min_height;

  my $child = $self->get_child;
  if ($child && $child->visible) {
    my $creq = $child->size_request;

    my $width_factor = $self->{'shrink_width_factor'};
    if ($width_factor > 0 && $creq->width * $width_factor <= $min_width) {
      $min_width = $creq->width;
    } else {
      $min_width = max ($min_width, $creq->width);
    }

    my $height_factor = $self->{'shrink_height_factor'};
    if ($height_factor >0 && $creq->height * $height_factor <= $min_height) {
      $min_height = $creq->height;
    } else {
      $min_height = max ($min_height, $creq->height);
    }
  }

  if (DEBUG) {
    if ($min_width  != $old_min_width || $min_height != $old_min_height) {
      print $self->get_name," request ",$min_width,"x",$min_height,
        ", extending min ",$old_min_width,"x",$old_min_height;
      if ($child) {
        my $creq = $child->size_request;
        print ", for child ", ($child->visible ? '' : '(not visible) '),
          "req ",$creq->width,"x",$creq->height;
      }
      print " border ",$self->get_border_width,"\n";
    }
  }

  # set and notify any new minimum for the width/height from the child
  #
  $self->{'minimum_width'}  = $min_width;
  $self->{'minimum_height'} = $min_height;
  # believe cleanest to notify after both width and height updated
  if ($old_min_width  != $min_width)  { $self->notify ('minimum-width'); }
  if ($old_min_height != $min_height) { $self->notify ('minimum-height'); }

  my $border_width = $self->get_border_width;
  $req->width  ($min_width  + 2*$border_width);
  $req->height ($min_height + 2*$border_width);
}

# 'size-allocate' class closure
#
# called by our parent to give us actual allocated space -- pass this down
# to the child, less the border width
# 
sub do_size_allocate {
  my ($self, $alloc) = @_;
  if (my $child = $self->get_child) {
    my $border_width  = $self->get_border_width;
    my $x = $alloc->x + $border_width;
    my $y = $alloc->y + $border_width;
    my $width = max (1, $alloc->width  - 2*$border_width);
    my $height = max (1, $alloc->height - 2*$border_width);

    my $child_alloc = $child->allocation;

    if (DEBUG) {
      my $creq = $child->size_request;
      print "NoShrink child alloc ${width}x${height} at $x,$y",
        ", vs child req ",$creq->width,"x",$creq->height,
          ", and current child ",
            $child_alloc->x,",",$child_alloc->y,
              " ",$child_alloc->width,"x",$child_alloc->height,
                "\n";
    }
    if ($x != $child_alloc->x
        || $y != $child_alloc->y
        || $width != $child_alloc->width
        || $height != $child_alloc->height) {
      $child->size_allocate (Gtk2::Gdk::Rectangle->new ($x, $y, $width, $height));
    }
  }
}

1;
__END__

=head1 NAME

Gtk2::Ex::NoShrink -- non-shrinking container widget

=for test_synopsis my ($my_child_widget)

=head1 SYNOPSIS

 use Gtk2::Ex::NoShrink;
 my $noshrink = Gtk2::Ex::NoShrink->new;
 $noshrink->add ($my_child_widget);

=head1 WIDGET HIERARCHY

C<Gtk2::Ex::NoShrink> is a subclass of C<Gtk2::Bin>.

    Gtk2::Widget
      Gtk2::Container
        Gtk2::Bin
          Gtk2::Ex::NoShrink

=head1 DESCRIPTION

A C<Gtk2::Ex::NoShrink> container widget holds a single child widget and
imposes a "no shrink" policy on its size.  The child can grow, but any
request to shrink is ignored.

When the child requests a size the NoShrink sets that as the size it
requests in turn from its parent.  If the child later changes asks to be
smaller, the NoShrink stays at the previous larger size, thus keeping the
child's largest-ever request.  A largest size is maintained separately for
width and for height.

Requested sizes are of course just that: only requests.  It's a matter for
the NoShrink's parent how much space is actually provided.  The NoShink
always sets its child to the full allocated space, less the usual
C<border-width> (see L<Gtk2::Container>) if that's set.  As usual it's then
a matter for the child what it does in a size perhaps bigger or perhaps
smaller than what it said it wanted.

If a child is added but not shown (no C<< $child->show >>) then it's treated
as if there was no child.  This is the same as other container classes do.
For NoShrink it means a size request of the C<minimum-width> by
C<minimum-height>, plus C<border-width>, and nothing drawn.

=head1 FUNCTIONS

=over 4

=item C<< Gtk2::Ex::NoShrink->new (key=>value,...) >>

Create and return a new C<Gtk2::Ex::NoShrink> container widget.  Optional
key/value pairs set initial properties per C<< Glib::Object->new >>.

    my $noshrink = Gtk2::Ex::NoShrink->new (border_width => 4);

The child widget can be set with the usual container C<add> method,

    $noshrink->add ($my_child_widget);

Or with the usual C<child> pseudo-property, either at creation time or
later,

    my $noshrink = Gtk2::Ex::NoShrink->new
                     (child => $my_child_widget);

=back

=head1 PROPERTIES

=over 4

=item C<minimum-width> (integer, default 0)

=item C<minimum-height> (integer, default 0)

The currently accumulated minimum width and height to impose on the size
request made by the child.  These are maintained as the largest width and
height requested from the child so far, or 0 for no size requested in that
direction so far.

Both sizes are for the child's space.  Any C<border-width> amount is added
on top of these for the size requested in turn by the NoShrink to its
parent.

These minimums can be set initially to begin at a particular size,

    my $noshrink = Gtk2::Ex::NoShrink->new (minimum_width => 30,
                                            minimum_height => 20);

They can also be reduced later to reset the size.  This results in the
greater of the new value or the child's current requested size.  Setting 0
for instance always goes to the child's current requested size.  This is
good if the nature of the child's content has changed,

    $noshrink->set('minimum-width', 0);  # reset to child size

These minimum size properties are unaffected by any removal of the child
widget or re-add of a new child.  If a new widget has a completely different
nature then you may want to reset both dimensions to start again from its
new sizes.

=item C<shrink-width-factor> (float, default 0)

=item C<shrink-height-factor> (float, default 0)

If non-zero then these values are factors representing a point at which the
NoShrink will in fact shrink to match the child's size.

If the child asks for a width which is a factor C<shrink-width-factor> (or
more) smaller than the existing C<minimum-width> imposed, then the
C<NoShrink> will reset the C<minimum-width> to the child's request.
Likewise C<shrink-height-factor> on the height.

Suppose for instance C<shrink-width-factor> is 2.0 and the NoShrink has over
the course of successive child requests grown to 300 pixels wide.  If the
child then asks to be 100 pixels the NoShrink will obey that, because it's a
shrinkage by a factor more than 2.0 times.  The new C<minimum-width> becomes
100.

These factors allow small shrinks to be ignored but large ones obeyed.  This
is good if the child might grow to a silly size when displaying wild data
for a while and you'd like it to shrink back when normality is restored.

=back

The usual C<border-width> property from C<Gtk2::Container> is followed.  It
leaves a border of that many pixels around the child in the C<NoShrink>
container's allocated space.  Currently nothing is drawn in the border.  The
default C<border-width> is 0, giving the child all the allocated space.

=head1 SEE ALSO

L<Gtk2::Container>, L<Gtk2::Bin>

=head1 HOME PAGE

L<http://user42.tuxfamily.org/gtk2-ex-noshrink/index.html>

=head1 LICENSE

Gtk2-Ex-NoShrink is Copyright 2007, 2008, 2009, 2010 Kevin Ryde

Gtk2-Ex-NoShrink is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation; either version 3, or (at your option) any later
version.

Gtk2-Ex-NoShrink is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
more details.

You should have received a copy of the GNU General Public License along with
Gtk2-Ex-NoShrink.  If not, see L<http://www.gnu.org/licenses/>.

=cut
